### PR TITLE
[pytx]  Implement a new cleaner PDQ index solution 

### PR DESCRIPTION
**Pull request by haianhng31**
Summary
---------
Resolve issue #1613 . 
This PR introduces a new SignalTypeIndex2 class for managing and querying a PDQ hash-based index using FAISS. Key changes include:
* [**_PDQHashIndex**](https://github.com/facebook/ThreatExchange/pull/1695/files#diff-2173010ed470e9f1a0daa6c696a604eb0937458100d74be1b326ab11ed3224d0R20-R98): A wrapper around the FAISS index to handle the serialization/deserialization of PDQ hashes, along with methods for adding hashes and performing searches using the FAISS library.
* **SignalTypeIndex2**: A class for managing a PDQ index, which includes:
  * The ability to add PDQ hashes and associate them with entries. ([lines](https://github.com/facebook/ThreatExchange/pull/1695/files#diff-2173010ed470e9f1a0daa6c696a604eb0937458100d74be1b326ab11ed3224d0R138-R153))
  * A query method for searching the index for matching entries based on a query hash. ([lines](https://github.com/facebook/ThreatExchange/pull/1695/files#diff-2173010ed470e9f1a0daa6c696a604eb0937458100d74be1b326ab11ed3224d0R128-R137))
  * A serialize and deserialize method for persisting and loading the index from binary streams using pickle. ([lines](https://github.com/facebook/ThreatExchange/pull/1695/files#diff-2173010ed470e9f1a0daa6c696a604eb0937458100d74be1b326ab11ed3224d0R155-R166))

Test Plan
---------
I have included several test cases for this, currently all in one file: 
* test for correct initialization 
* test serialize deserialize (with & without custom index)
* search functionality
  * test empty index query -> return empty array 
  * test sample with 1 exact match 
  * test sample with 1 near exact match 
  * test distance threshold behavior 
  * test duplicate handling

### CODE REVIEW

**Comment by Dcallies** [comment]
Sorry I didn't get to this today, will check first thing tomorrow!

**Review by Dcallies — CHANGES_REQUESTED** [review]
blocking: #1613 is about creating a new PDQ index implementation which follows the SignalType interface, but this PR creates a new SignalTypeIndex interface entirely, which is surprising.

Is this an intentional change, and if so, can you walk me through why we should should go this route than a more tailored approach to provide a second, cleaned up PDQ index implementation of the SignalTypeIndex interface which lives in https://github.com/facebook/ThreatExchange/tree/main/python-threatexchange/threatexchange/signal_type/pdq ?

As part of this, can you include tests that show that the bug in #1318 is solved?

**Comment by haianhng31** [comment]
I appreciate your feedback on this PR. To be honest, I was a bit uncareful at the start and didn't fully realize that creating a new SignalTypeIndex was changing the whole interface, rather than a more tailored approach.

But since now that is cleared up, moving forward, I will provide a second, cleaned up PDQ index implementation that adheres to the existing SignalTypeIndex interface. I'll be sure to include needed tests as well!

File: python-threatexchange/threatexchange/signal_type/index2.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+            threshold (int): The maximum distance threshold for matches.
+            faiss_index (faiss.Index): An optional pre-existing FAISS index to use.
+        """
+        super().__init__()
+        if faiss_index is None:
+            # Use a simple brute-force FAISS index by default
+            faiss_index = faiss.IndexFlatL2(DIMENSIONALITY)
+        self.faiss_index = _PDQHashIndex(faiss_index)
+        self.threshold = threshold
+        self._deduper: t.Set[str] = set()
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/index2.py:125]: From the meeting:

Should store this as a mapping from hash => faiss_id 

add(h, payload1)
add(h, playload2)

deduper: dict[hash, faiss_id] = dict[]
entries: list[int, payload] = list[]

existing_id = deduper.get(new_hash)
if existing_id is not None: 
   # Don't add to faiss!
   # we add to our internal entry mapping
   entries[existing_id].append(payload2)
else:
  # faiss id is 0 -> size
  next_id = len(deduper)
  faiss.add(h)
  entries.append([payload])
  deduper[h] = next_id

///
lookup(h) -> payloads
  faiss_id = faiss.search(h) 
  entries[faiss_id]

query(h) -> [payload1, payload2]
```

**Pull request closed without merging**